source/UF/Truncations.lagda

```diff
@@ -0,0 +1,435 @@
+ canonical-pred-map {𝓤} {X} {n} x =
+  ∥∥ₙ-rec (hlevels-are-upper-closed n (∥ X ∥[ n ]) (∥∥ₙ-hlevel n))
+          (λ x → ∣ x ∣[ n ]) x
+
+ canonical-pred-map-comp : {X : 𝓤 ̇} {n : ℕ} (x : X)
+                         → canonical-pred-map (∣ x ∣[ succ n ]) ＝ (∣ x ∣[ n ])
+ canonical-pred-map-comp {𝓤} {X} {n} x =
+  ∥∥ₙ-rec-comp (hlevels-are-upper-closed n (∥ X ∥[ n ]) (∥∥ₙ-hlevel n))
+               (λ _ → ∣ _ ∣[ n ]) x
+
+ to-∼-of-maps-with-truncated-domain : {n : ℕ} {X : 𝓤 ̇} {Y : 𝓥 ̇}
+                                     → (f g : ∥ X ∥[ n ] → Y)
+                                     → Y is-of-hlevel n
+                                     → ((x : X)
+                                      → f (∣ x ∣[ n ]) ＝ g (∣ x ∣[ n ]))
+                                     → f ∼ g
+ to-∼-of-maps-with-truncated-domain {𝓤} {𝓥} {n} {X} {Y} f g Y-hlev H =
+  ∥∥ₙ-ind (λ - → hlevels-are-closed-under-id n Y-hlev (f -) (g -)) H
+
+ to-∼-of-maps-between-truncated-types : {n : ℕ} {X : 𝓤 ̇} {Y : 𝓥 ̇}
+                                      → (f g : ∥ X ∥[ n ] → ∥ Y ∥[ n ])
+                                      → ((x : X)
+                                       → f (∣ x ∣[ n ]) ＝ g (∣ x ∣[ n ]))
+                                      → f ∼ g
+ to-∼-of-maps-between-truncated-types {𝓤} {𝓥} {n} {X} {Y} f g H =
+  to-∼-of-maps-with-truncated-domain f g (∥∥ₙ-hlevel n) H
+
+ truncation-closed-under-equiv : {n : ℕ} {X : 𝓤 ̇} {Y : 𝓥 ̇}
+                               → X ≃ Y
+                               → ∥ X ∥[ n ] ≃ ∥ Y ∥[ n ]
+ truncation-closed-under-equiv {𝓤} {𝓥} {n} {X} {Y} e = (f , (b , G) , (b , H))
+  where
+   f : ∥ X ∥[ n ] → ∥ Y ∥[ n ]
+   f = ∥∥ₙ-rec (∥∥ₙ-hlevel n) (λ x → ∣ ⌜ e ⌝ x ∣[ n ])
+   b : ∥ Y ∥[ n ] → ∥ X ∥[ n ]
+   b = ∥∥ₙ-rec (∥∥ₙ-hlevel n) (λ y → ∣ ⌜ e ⌝⁻¹ y ∣[ n ])
+   H : b ∘ f ∼ id
+   H = to-∼-of-maps-between-truncated-types (b ∘ f) id H'
+    where
+     H' : (x : X) → b (f (∣ x ∣[ n ])) ＝ (∣ x ∣[ n ])
+     H' x = b (f (∣ x ∣[ n ]))           ＝⟨ I ⟩
+            b (∣ ⌜ e ⌝ x ∣[ n ])         ＝⟨ II ⟩
+            (∣ ⌜ e ⌝⁻¹ (⌜ e ⌝ x) ∣[ n ]) ＝⟨ III ⟩
+            (∣ x ∣[ n ])                 ∎
+      where
+       I = ap b (∥∥ₙ-rec-comp (∥∥ₙ-hlevel n) (λ x → ∣ (⌜ e ⌝ x) ∣[ n ]) x)
+       II = ∥∥ₙ-rec-comp (∥∥ₙ-hlevel n) (λ y → ∣ (⌜ e ⌝⁻¹ y) ∣[ n ]) (⌜ e ⌝ x)
+       III = ap (λ x → ∣ x ∣[ n ]) (inverses-are-retractions' e x)
+   G : f ∘ b ∼ id
+   G = to-∼-of-maps-between-truncated-types (f ∘ b) id G'
+    where
+     G' : (y : Y) → f (b (∣ y ∣[ n ])) ＝ (∣ y ∣[ n ])
+     G' y = f (b (∣ y ∣[ n ]))           ＝⟨ I ⟩
+            f (∣ (⌜ e ⌝⁻¹ y) ∣[ n ])     ＝⟨ II ⟩
+            (∣ ⌜ e ⌝ (⌜ e ⌝⁻¹ y) ∣[ n ]) ＝⟨ III ⟩
+            (∣ y ∣[ n ])                 ∎
+      where
+       I = ap f (∥∥ₙ-rec-comp (∥∥ₙ-hlevel n) (λ y → ∣ (⌜ e ⌝⁻¹ y) ∣[ n ]) y)
+       II = ∥∥ₙ-rec-comp (∥∥ₙ-hlevel n) (λ x → ∣ ⌜ e ⌝ x ∣[ n ]) (⌜ e ⌝⁻¹ y)
+       III = ap (λ y → ∣ y ∣[ n ]) (inverses-are-sections' e y)
+
+ successive-truncations-equiv : (X : 𝓤 ̇) (n : ℕ)
```

Comment:
The arguments `X` and `n` are implicit in (almost?) all your other constructions. Should they be implicit here too?